plugins/modules/katello_content_view_version.py

```diff
@@ -143,58 +127,34 @@
     state: absent
 '''
 
-RETURN = '''# '''
+RETURN = ''' # '''
 
-try:
-    from nailgun.entities import (
-        ContentViewVersion,
-    )
 
-    from ansible.module_utils.ansible_nailgun_cement import (
-        create_server,
-        find_organization,
-        find_lifecycle_environments,
-        find_lifecycle_environment,
-        find_content_view,
-        find_content_view_version,
-        ping_server,
-        set_task_timeout,
-        naildown_entity_state,
-    )
-    has_import_error = False
-except ImportError as e:
-    has_import_error = True
-    import_error_msg = str(e)
-from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.foreman_helper import KatelloEntityAnsibleModule
 
 
-def promote_content_view_version(module, content_view_version, organization, environments, synchronous, **kwargs):
+def promote_content_view_version(module, content_view_version, environments, synchronous, force, force_yum_metadata_regeneration):
     changed = False
 
-    current_environment_ids = map(lambda environment: environment.id, content_view_version.environment)
-    desired_environment_ids = map(lambda environment: environment.id, environments)
-    promote_to_environment_ids = list(set(desired_environment_ids) - set(current_environment_ids))
-
-    request_data = {'environment_ids': promote_to_environment_ids}
-    request_data.update({k: v for k, v in kwargs.items() if v is not None})
+    current_environment_ids = {environment['id'] for environment in content_view_version['environments']}
+    desired_environment_ids = {environment['id'] for environment in environments}
+    promote_to_environment_ids = list(desired_environment_ids - current_environment_ids)
 
     if promote_to_environment_ids:
-        if not module.check_mode:
-            content_view_version.promote(synchronous, data=request_data)
-        changed = True
+        payload = {
+            'environment_ids': promote_to_environment_ids,
+            'force': force,
+            'force_yum_metadata_regeneration': force_yum_metadata_regeneration,
+        }
+
```

Comment:
Seems promote_content_view_version doesn't pass synchronous to the resource action?

Reply:
Bonus points if you can add a test like 

https://github.com/theforeman/foreman-ansible-modules/blob/3a85d43dffb4f93daf791d05069862d1c8b743e0/tests/test_playbooks/tasks/katello_sync.yml#L21-L28

Reply:
There is a problem here. I think we can only trigger the promote task, once the publish task is done. So we cannot have them both synchronous. What should we do about it?
Can we fold the promotion into the publish request? Can we add a wait for _this_ task to the promote request?

Reply:
I think we can have them both sync, just not both unsync?

We issue the publish here:
https://github.com/theforeman/foreman-ansible-modules/blob/8d57b2f211201d9b0cc46d7245378a3dfab3554c/plugins/modules/katello_content_view_version.py#L209

If sync is `True`, this step will wait until the publish is done and we can do a sync promote too.

If sync=False for the publish, then already this line won't work (I think, didn't verify):
https://github.com/theforeman/foreman-ansible-modules/blob/8d57b2f211201d9b0cc46d7245378a3dfab3554c/plugins/modules/katello_content_view_version.py#L210

And we can't promote until the publish is done.

IMHO we should do the following:
* no current_lifecycle_environment, no version and sync=False → error out during param validation
* all other cases look happy?